### PR TITLE
Time based BDA averaging

### PR DESCRIPTION
**Pull request by smasoka**
- [ ] Tests added / passed

  ```bash
  $ py.test -v -s africanus
  ```

  If the pep8 tests fail, the quickest way to correct
  this is to run `autopep8` and then `flake8` and
  `pycodestyle` to fix the remaining issues.

  ```
  $ pip install -U autopep8 flake8 pycodestyle
  $ autopep8 -r -i africanus
  $ flake8 africanus
  $ pycodestyle africanus
  ```

- [ ] Fully documented, including `HISTORY.rst` for all changes
      and one of the `docs/*-api.rst` files for new API

  To build the docs locally:

  ```
  pip install -r requirements.readthedocs.txt
  cd docs
  READTHEDOCS=True make html
  ```

### CODE REVIEW

**Comment by sjperkins** [comment]
delta-uvw/delta-time code here:

https://github.com/ska-sa/codex-africanus/pull/117/files#diff-b031c4618503b262e57d1b435bd50bc7R50

As discussued I think you should be passing in the UVW coordinate array and binning samples based on how far a UVW track has travelled.

File: small_ms_test_case_baseline.py [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+        if r == -1:
+            continue
+
+            
+#         half_int = interval[r] * 0.5
+#         print("half_int",half_int)
+#         if bin_count == 0:
+#             bin_low = timestamps[r] - half_int
+#             print("bin_count ", bin_count)
+#             print("bin_low", bin_low)
```
Comment (sjperkins) [inline small_ms_test_case_baseline.py]: At this point, there's nothing in our current bin ``bin_count == 0`` so we establish a lower interval on samples that we will add to the new bin. For time averaging, this is time minus half the integration period.

For UVW averaging we're including samples based on the UV distance travelled, so this low point is going to be some UVW coordinate or UVW distance travelled measure.

File: small_ms_test_case_baseline.py [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+#         elif timestamps[r] + half_int - bin_low > bl_dist_time_bins[bl]:
```
Comment (sjperkins) [inline small_ms_test_case_baseline.py]: At this point, we're adding samples to our current bin. For time averaging we test if the current time plus half integration period minus ``bin_low`` is less than some measure. If so, we add the sample. If not we finish the bin and create a new one.

In the UVW averaging case, the test will probably involve checking whether we've travelled far enough in UVW space in order to create a new bin. If we haven't we can safely add the current sample to the bin.

**Comment by smasoka** [comment]
This is the error I get running `test_baseline_time_and_channel.py`

```
(base) Sakhiles-MacBook-Pro:tests smasoka$ python test_baseline_time_and_channel.py 
Running test_baseline_time_and_channel
Traceback (most recent call last):
  File "test_baseline_time_and_channel.py", line 55, in <module>
    bins_for_longest_baseline=1.0)
  File "/anaconda3/lib/python3.7/site-packages/numba/dispatcher.py", line 348, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/anaconda3/lib/python3.7/site-packages/numba/dispatcher.py", line 315, in error_rewrite
    reraise(type(e), e, None)
  File "/anaconda3/lib/python3.7/site-packages/numba/six.py", line 658, in reraise
    raise value.with_traceback(tb)
numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
Internal error at <numba.typeinfer.StaticGetItemConstraint object at 0x616e8a668>:
--%<----------------------------------------------------------------------------
Traceback (most recent call last):
  File "/anaconda3/lib/python3.7/site-packages/numba/errors.py", line 609, in new_error_context
    yield
  File "/anaconda3/lib/python3.7/site-packages/numba/typeinfer.py", line 340, in __call__
    value=ty, index=self.index)
  File "/anaconda3/lib/python3.7/site-packages/numba/typing/context.py", line 299, in resolve_static_getitem
    return self.resolve_function_type("static_getitem", args, kws)
  File "/anaconda3/lib/python3.7/site-packages/numba/typing/context.py", line 203, in resolve_function_type
    res = self._resolve_builtin_function_type(func, args, kws)
  File "/anaconda3/lib/python3.7/site-packages/numba/typing/context.py", line 220, in _resolve_builtin_function_type
    res = defn.apply(args, kws)
  File "/anaconda3/lib/python3.7/site-packages/numba/typing/templates.py", line 206, in apply
    sig = generic(args, kws)
  File "/anaconda3/lib/python3.7/site-packages/numba/typing/builtins.py", line 613, in generic
    return tup.types[idx]
IndexError: tuple index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/anaconda3/lib/python3.7/site-packages/numba/typeinfer.py", line 141, in propagate
    constraint(typeinfer)
  File "/anaconda3/lib/python3.7/site-packages/numba/typeinfer.py", line 345, in __call__
    self.fallback(typeinfer)
  File "/anaconda3/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/anaconda3/lib/python3.7/site-packages/numba/errors.py", line 617, in new_error_context
    six.reraise(type(newerr), newerr, tb)
  File "/anaconda3/lib/python3.7/site-packages/numba/six.py", line 659, in reraise
    raise value
numba.errors.InternalError: tuple index out of range
[1] During: typing of static-get-item at /Users/smasoka/Varsity/codex-africanus/africanus/averaging/baseline_time_and_channel_avg.py (64)
--%<----------------------------------------------------------------------------


File "../baseline_time_and_channel_avg.py", line 64:
        def impl(output, orow, input, irow):
            for c in range(output.shape[1]):
            ^

[1] During: resolving callee type: type(CPUDispatcher(<function comp_add_factory.<locals>.impl at 0x616e75378>))
[2] During: typing of call at /Users/smasoka/Varsity/codex-africanus/africanus/averaging/baseline_time_and_channel_avg.py (147)


File "../baseline_time_and_channel_avg.py", line 147:
    def impl(meta, ant1, ant2, interval=None, flag_row=None, time_centroid=None, 
        <source elided>
            if flags_match(flag_row, in_row, meta.flag_row, out_row):
                uvw_adder(interval_avg, out_row, interval, in_row)
                ^

[1] During: resolving callee type: type(CPUDispatcher(<function baseline_row_average at 0x114977e18>))
[2] During: typing of call at /Users/smasoka/Varsity/codex-africanus/africanus/averaging/baseline_time_and_channel_avg.py (282)


File "../baseline_time_and_channel_avg.py", line 282:
    def impl(time, interval, antenna1, antenna2,
        <source elided>
        # Average the rows according to the meta data
        row_data = baseline_row_average(row_meta, antenna1, antenna2, interval, flag_row=flag_row,                                                  time_centroid=time_centroid, exposure=exposure, uvw=uvw, weight=weight, sigma=sigma)
        ^

This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.

To see Python/NumPy features supported by the latest release of Numba visit:
http://numba.pydata.org/numba-doc/dev/reference/pysupported.html
and
http://numba.pydata.org/numba-doc/dev/reference/numpysupported.html

For more information about typing errors and how to debug them visit:
http://numba.pydata.org/numba-doc/latest/user/troubleshoot.html#my-code-doesn-t-compile

If you think your code should work with Numba, please report the error message
and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new

(base) Sakhiles-MacBook-Pro:tests smasoka$
```

**Comment by sjperkins** [comment]
@smasoka uvw_adder expects a ('row', 3) shape, but you're passing a ('row',) interval column

**Comment by smasoka** [comment]
```
(base) Sakhiles-MacBook-Pro:tests smasoka$ python test_baseline_time_and_channel.py 
Running test_baseline_time_and_channel
Traceback (most recent call last):
  File "test_baseline_time_and_channel.py", line 55, in <module>
    bins_for_longest_baseline=1.0)
  File "/Users/smasoka/Varsity/codex-africanus/africanus/averaging/baseline_time_and_channel_avg.py", line 205, in impl
    raise ValueError("flag_row and flag arrays mismatch")
ValueError: flag_row and flag arrays mismatch
(base) Sakhiles-MacBook-Pro:tests smasoka$ 
```

**Comment by sjperkins** [comment]
FLAG is a high-resolution version of FLAG_ROW. The two need to agree. This can be achieved with

```python
flag_row = np.any(flag, axis=(1, 2)
```

File: africanus/averaging/baseline_time_and_channel_mapping.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        
+        # Unique baseline and time 
+        ubl, bl_idx, bl_inv, bl_count = unique_baselines(antenna1, antenna2)
+        utime, _, time_inv, _ = unique_time(time)
+        
+        # Calculate distances of the unique baselines
+        for i in range(ubl.shape[0]): 
+            p,q = ubl[i,0],ubl[i,1]
+            uvwpq = uvw[(antenna1 == p) & (antenna2 == q)] 
+            baseline_dist = np.sqrt((uvwpq**2).sum(axis=1))[0]
```
Comment (sjperkins) [inline africanus/averaging/baseline_time_and_channel_mapping.py]: @atemkeng This seems to set the baseline distance to the UVW distance of the first selected row, for the baseline.

Is this accurate and can it be improved upon? /cc @smasoka

**Comment by smasoka** [comment]
latest commit "Testing using MS"

Currently, running bda from the bda_dist_pilot 

```
(base) Sakhiles-MacBook-Pro:tests smasoka$ time python ~/Varsity/bda-dist-pilot/bda_dist_pilot/bda_time_interface-2.py ~/Varsity/Sample-MeerKATDATA/MeerKATDATA/1491291289.1GC.ms
Successful read/write open of default-locked table /Users/smasoka/Varsity/Sample-MeerKATDATA/MeerKATDATA/1491291289.1GC.ms: 26 columns, 129480 rows
Number of Antennas
15
Number of Baselines
105.0
Desc
(129480,)
UVW
(129480, 3)
Flag
(129480, 119, 2)
Flag Row
(129480,)
Antenna 1
(129480,)
Antenna 2
(129480,)
Interval
(129480,)
Exposure
(129480,)
Time
(129480,)
Time Centroid
(129480,)
DATA
(129480, 119, 2)
Weights
(129480, 2)
Scan
(129480,)
------------------
Compressed Arrays
------------------
Desc Output
(66927,)
UVW Output
(66927, 3)
Flag Output
(66927, 119, 2)
Flag Row Output
(66927,)
Antenna 1 Output
(66927,)
Antenna 2 Output
(66927,)
Interval Output
(66927,)
Exposure Output
(66927,)
Time Output
(66927,)
DATA Output
(66927, 119, 2)
Time Centroid Output
(66927,)
Weights Output
(66927, 2)
Scan Output
(66927,)

real	2m57.611s
user	6m0.761s
sys	5m27.637s
(base) Sakhiles-MacBook-Pro:tests smasoka$ 
```

and Running bda averaging from africanus

```
(base) Sakhiles-MacBook-Pro:tests smasoka$ time python test_baseline_time_and_channel-MS.py ~/Varsity/Sample-MeerKATDATA/MeerKATDATA/1491291289.1GC.ms
Successful read/write open of default-locked table /Users/smasoka/Varsity/Sample-MeerKATDATA/MeerKATDATA/1491291289.1GC.ms: 26 columns, 129480 rows
Original time 
 (129480,)
Orginal interval 
 (129480,)
Orginal antenna1 
 (129480,)
Orginal antenna2 (129480,)
Orginal uvw 
 (129480, 3)
Orginal weight 
 (129480, 2)
Orginal sigma 
 (129480, 2)
Running test_baseline_time_and_channel
In baseline time and channel averaging
65778
(129480, 3)
avg.time 
 (65778,)
avg.interval 
 (65778,)
avg.antenna1 
 (65778,)
avg.antenna2 (65778,)
avg.uvw 
 (65778, 3)
avg.weight 
 (65778, 2)
avg.sigma 
 (65778, 2)

real	0m3.953s
user	0m3.839s
sys	0m0.446s
(base) Sakhiles-MacBook-Pro:tests smasoka$
```

I'm expecting the number of rows compressed to be the same since the logic of the code is the same `longest_baseline_distance//current_baseline_distance` but its not, so there is some accuracy I need to be working on. This is just baseline time, I haven't started on baseline channel since there are a few things that need improvement on baseline time.

File: africanus/averaging/baseline_time_and_channel_mapping.py (hunk 3 above)
Comment (sjperkins) [reply africanus/averaging/baseline_time_and_channel_mapping.py]: So would it be more accurate to do?

```python
baseline_dist = np.sqrt((uvwpq**2).sum()
```

Comment (smasoka) [reply africanus/averaging/baseline_time_and_channel_mapping.py]: `baseline_dist = np.abs(uvwpq[:, 0]).sum(axis=0)`
using the u coordinates only

File: africanus/averaging/baseline_time_and_channel_avg.py [hunk 4]
```diff
@@ -0,0 +1,575 @@
...
+def chan_output_factory(present):
+    """ Returns function producing outputs if the array is present """
+    if present:
+        def impl(shape, array):
+            return np.zeros(shape, dtype=array.dtype)
+    else:
+        def impl(shape, array):
+            pass
+
+    return njit(nogil=True, cache=True)(impl)
```
Comment (smasoka) [inline africanus/averaging/baseline_time_and_channel_avg.py]: @sjperkins can you help me look into this. I get numba compilation error on the 'chan_output_factory' function

```
Traceback (most recent call last):
  File "test_baseline_time_and_channel.py", line 76, in <module>
    bins_for_longest_baseline=1.0, baseline_chan_bin_size=1)
  File "/anaconda3/lib/python3.7/site-packages/numba/dispatcher.py", line 376, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/anaconda3/lib/python3.7/site-packages/numba/dispatcher.py", line 343, in error_rewrite
    reraise(type(e), e, None)
  File "/anaconda3/lib/python3.7/site-packages/numba/six.py", line 658, in reraise
    raise value.with_traceback(tb)
numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<built-in function zeros>) with argument(s) of type(s): ((int64, array(uint32, 1d, C), int64), dtype=dtype(complex64))
 * parameterized
In definition 0:
    All templates rejected with literals.
In definition 1:
    All templates rejected without literals.
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<built-in function zeros>)
[2] During: typing of call at /Users/smasoka/Varsity/codex-africanus/africanus/averaging/baseline_time_and_channel_avg.py (194)


File "../baseline_time_and_channel_avg.py", line 194:
        def impl(shape, array):
            return np.zeros(shape, dtype=array.dtype)
            ^

[1] During: resolving callee type: type(CPUDispatcher(<function chan_output_factory.<locals>.impl at 0x619b0abf8>))
[2] During: typing of call at /Users/smasoka/Varsity/codex-africanus/africanus/averaging/baseline_time_and_channel_avg.py (348)

[3] During: resolving callee type: type(CPUDispatcher(<function chan_output_factory.<locals>.impl at 0x619b0abf8>))
[4] During: typing of call at /Users/smasoka/Varsity/codex-africanus/africanus/averaging/baseline_time_and_channel_avg.py (353)

[5] During: resolving callee type: type(CPUDispatcher(<function chan_output_factory.<locals>.impl at 0x619b0abf8>))
[6] During: typing of call at /Users/smasoka/Varsity/codex-africanus/africanus/averaging/baseline_time_and_channel_avg.py (348)


File "../baseline_time_and_channel_avg.py", line 348:
    def impl(row_meta, chan_meta,flag_row=None, vis=None, flag=None,
        <source elided>

        vis_avg = vis_factory(out_shape, vis)
        ^

[1] During: resolving callee type: type(CPUDispatcher(<function baseline_chan_average at 0x11927d730>))
[2] During: typing of call at /Users/smasoka/Varsity/codex-africanus/africanus/averaging/baseline_time_and_channel_avg.py (558)

[3] During: resolving callee type: type(CPUDispatcher(<function baseline_chan_average at 0x11927d730>))
[4] During: typing of call at /Users/smasoka/Varsity/codex-africanus/africanus/averaging/baseline_time_and_channel_avg.py (558)


File "../baseline_time_and_channel_avg.py", line 558:
    def impl(time, interval, antenna1, antenna2,
        <source elided>
                                          flag=flag, weight_spectrum=weight_spectrum, sigma_spectrum=sigma_spectrum,
                                          baseline_chan_bin_size=baseline_chan_bin_size)
                                          ^

This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.

To see Python/NumPy features supported by the latest release of Numba visit:
http://numba.pydata.org/numba-doc/latest/reference/pysupported.html
and
http://numba.pydata.org/numba-doc/latest/reference/numpysupported.html

For more information about typing errors and how to debug them visit:
http://numba.pydata.org/numba-doc/latest/user/troubleshoot.html#my-code-doesn-t-compile

If you think your code should work with Numba, please report the error message
and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new
```

Comment (smasoka) [reply africanus/averaging/baseline_time_and_channel_avg.py]: I tried to reproduce to try understand but

```
In [2]: import numpy as np 
   ...: from numba import njit 
   ...: from numba import generated_jit 
   ...:  
   ...: nchan = 16 
   ...: ncorr = 4 
   ...: vis = (np.arange(10*nchan*ncorr, dtype=np.float32) + 
   ...:       np.arange(1, 10*nchan*ncorr+1, dtype=np.float32)*1j) 
   ...: vis = vis.reshape(10, nchan, ncorr) 
   ...:  
   ...: shape = (10, nchan, ncorr) 
   ...: np.zeros(shape, dtype=vis.dtype) 
   ...:  
   ...: @generated_jit(nopython=True, nogil=True, cache=True) 
   ...: def call_impl(shape, vis): 
   ...:     impl(shape, vis) 
   ...:  
   ...: @njit(nogil=True, cache=True) 
   ...: def impl(shape, array): 
   ...:     return np.zeros(shape, dtype=array.dtype) 
   ...:                                                                                                                                                       

In [3]: call_impl(shape,vis)                                                                                                                                  
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
<ipython-input-3-25a1767fcd6c> in <module>
----> 1 call_impl(shape,vis)

/anaconda3/lib/python3.7/site-packages/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    374                 e.patch_message(msg)
    375 
--> 376             error_rewrite(e, 'typing')
    377         except errors.UnsupportedError as e:
    378             # Something unsupported is present in the user code, add help info

/anaconda3/lib/python3.7/site-packages/numba/dispatcher.py in error_rewrite(e, issue_type)
    341                 raise e
    342             else:
--> 343                 reraise(type(e), e, None)
    344 
    345         argtypes = []

/anaconda3/lib/python3.7/site-packages/numba/six.py in reraise(tp, value, tb)
    656             value = tp()
    657         if value.__traceback__ is not tb:
--> 658             raise value.with_traceback(tb)
    659         raise value
    660 

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Unknown attribute 'dtype' of type typeref[array(complex64, 3d, C)]

File "<ipython-input-2-f6e53a862b5e>", line 20:
def impl(shape, array):
    return np.zeros(shape, dtype=array.dtype)
    ^

[1] During: typing of get attribute at <ipython-input-2-f6e53a862b5e> (20)

File "<ipython-input-2-f6e53a862b5e>", line 20:
def impl(shape, array):
    return np.zeros(shape, dtype=array.dtype)
    ^

This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.
.......
```

**Comment by sjperkins** [comment]
generated_jit is not being used correctly. It is supposed to return a function defining an implementation for the types. Perhaps take a look at the [docs](https://numba.pydata.org/numba-doc/dev/user/generated-jit.html)?

The following works for me:

```python
import numpy as np
from numba import njit
from numba import generated_jit

nchan = 16
ncorr = 4
vis = (np.arange(10*nchan*ncorr, dtype=np.float32) +
      np.arange(1, 10*nchan*ncorr+1, dtype=np.float32)*1j)
vis = vis.reshape(10, nchan, ncorr)

shape = (10, nchan, ncorr)
np.zeros(shape, dtype=vis.dtype)

@generated_jit(nopython=True, nogil=True, cache=True)
def call_impl(shape, vis):
    def impl(shape, vis):
        return np.zeros(shape, dtype=vis.dtype)

    return impl


print(call_impl(shape, vis).shape)
```

**Pull request closed without merging**